common/metrics/otel_metrics_handler.go

```diff
@@ -37,13 +38,36 @@ import (
 )
 
 // otelMetricsHandler is an adapter around an OpenTelemetry [metric.Meter] that implements the [Handler] interface.
-type otelMetricsHandler struct {
-	l           log.Logger
-	tags        []Tag
-	provider    OpenTelemetryProvider
-	excludeTags excludeTags
-	catalog     catalog
-}
+type (
+	otelMetricsHandler struct {
+		l            log.Logger
+		set          attribute.Set
+		provider     OpenTelemetryProvider
+		tagConverter tagConverter
+		catalog      catalog
+		gauges       *sync.Map // string -> *gaugeAdapter. note: shared between multiple otelMetricsHandlers
+	}
+
+	tagConverter func(Tag) attribute.KeyValue
+
+	// This is to work around the lack of synchronous gauge:
+	// https://github.com/open-telemetry/opentelemetry-specification/issues/2318
+	gaugeAdapter struct {
+		lock   sync.Mutex
+		values map[attribute.Distinct]gaugeValue
+	}
+	gaugeValue struct {
+		value float64
+		// In practice, we can use attribute.Set itself as the map key in gaugeAdapter and it
+		// works, but according to the API we should use attribute.Distinct. But we can't get a
+		// Set back from a Distinct, so we have to store the set here also.
+		set attribute.Set
```

Comment:
Why do we need the set at all?

Reply:
to actually submit values to the observer